fixes #7258

jayesh92 opened the pull request:
This commit fixes #7258 
As per PY3 xrange is dropped and range is equivalent to PY2 xrange. 
As @asmeurer said in comments we should code as per PY3 and use
backward compatibility to serve PY2. I searched all files which were
using xrange without importing from compatibility (using grep -iR).
Only two were found and I have updated these files accordingly.
Please review this. All tests passed on my machine

smichr commented:
Thanks for working on this. Could you please squash these commits to a single commit and edit the opening paragraph of this PR by adding the line "fixes #7258". Although [the workflow](https://github.com/sympy/sympy/wiki/Development-workflow) (in "Other things to do in commit messages") suggests that you put the information in a commit message, I like to see it in the opening paragraph. You wrote "fix for #7258" which does not conform to the format given [here](https://help.github.com/articles/closing-issues-via-commit-messages/).

jayesh92 commented:
@smichr , I'm a beginner at git. I have updated the PR Title. I'm not sure on how to squash my last commits into one. Also my first commit was breaking the travis build, so I would only like to squash the last two commits into one.

smichr commented:
> I'm a beginner at git

Welcome to git :-) No problem. We have everyone from beginner to advanced working with SymPy. 

What happens when you squash all commits together is that the current state of the code base becomes the only thing recorded in the commit. So the fact that there was a failing commit will be erased with the squashing. Since you have not merged you can just reset back 3 commits with `git reset head~3`. You can use `git diff` to see the changes if you want. Now just commit those as if you just made those changes.

```
Or, if you like to work interactively, you can do `git rebase -i head~3` and 
you will be put into a vi session in which you can change the `pick` to things like 
`s` or `f` as described in that environment.  (And if you ever feel unsure about 
what you did during an interactive session you can just delete all the lines and 
exit with `:wq` and nothing will happen to the branch commits.) 
```

> I have updated the PR Title

I don't think it works with the title, that's why I directed you to the "opening paragraph" which is the first text box on this page. Click on the pencil in the upper right hand corner and enter the text then click the "Update Comment" button under the text box.

If you have any questions, feel free to ask.

asmeurer commented:
It does work with the title, but you should fix it anyway, because "fixes #123" isn't a very informative title, and because GitHub doesn't auto-link the issue number in the title.

jayesh92 commented:
@smichr I did the steps mentioned by you. I would just like to clarify on what I've essentially done in my PR. As per PY3 we should only use range which is basically the `xrange` of PY2. So in compatibility I've done `xrange = range` for PY3 and `xrange = xrange` for PY2. Also, I found that there were 2 files which were using `xrange` without importing from compatibility and so I imported `xrange` from compatibility in those 2 files, which would otherwise have broken the PY3 build.
Please correct me If I'm wrong ?

smichr commented:
@asmeurer , I was referring to the title of the PR. I checked the [github notes](https://help.github.com/articles/closing-issues-via-commit-messages/) and it only talks about using this syntax in commit messages. Since the title of a commit message using the syntax there should work as you noted. I also have found documentation about auto-closing [here](https://github.com/blog/1506-closing-issues-via-pull-requests) and it says that the syntax must appear in _the body_ of the PR. I'm not sure what that refers to: there are the _title_, _description_, _discussion_, and _merge_message_ areas. Do you know if the syntax works in all those areas? Is there are way to get the syntax to be ignored (e.g. by putting it in quotes, else poor old "issue #123" will be closed when this PR is committed since you and I both referred to it that way).

smichr commented:
> So in compatibility I've done xrange = range for PY3 and xrange = xrange for PY2

This is not showing up in your PR; only the imports of xrange from compatibility are showing.

jayesh92 commented:
@smichr This was done in my second commit, can be viewed at https://github.com/jayesh92/sympy/commit/dc247d7df942bc01acdbeecac569ef193209cd65

jayesh92 commented:
@smichr My understanding is that since PY3 does not has `xrange` so if some developer plans to use `xrange`  they better import it from compatibility else it wont work on PY3 however it would work on PY2 because PY2 provides `range` as well as `xrange`

asmeurer commented:
It will work because the title of the PR becomes the commit message of the merge commit. Although I've actually seen it work even when "fixes" was only in the PR OP message itself.

smichr commented:
> It will work because the title of the PR

OK, so as long as someone doesn't change the commit message, using it in the title will work, and in the OP. Do we know if it will work if it appears in the discussion area? Do we know if there is a way to escape such a reference (e.g. with quotes)?

asmeurer commented:
I doubt it works in the discussion, other than the OP message.

smichr commented:
> I doubt it works in the discussion

I just confirmed that writing fixes #foo  in the discussion does not close the issue when the PR is closed.

jayesh92 commented:
@asmeurer Could you please review my last commit. I've changed all files as you had explained.
Commit can be viewed at https://github.com/jayesh92/sympy/commit/a43b8309106b882c5d54bc6737dbcb3de84854b3

smichr commented on sympy/combinatorics/tests/test_perm_groups.py:
```diff
@@ -1,3 +1,4 @@
+from sympy.core.compatibility import tange
```
typo here: tange -> range

smichr commented:
Other than the one typo this looks good.

jayesh92 commented:
@smichr I've fixed the typo, I wonder how was it not checked in the tests ? Can this be merged, My PR is showing merge conflicts

smichr commented:
> showing merge conflicts

Yes. This means you should update your master branch and rebase this. And then before committing, produce a single commit. If you are there I could go to a chat page to talk you through the process. I forget which page sympy uses for chatting, however. Do you know?

jayesh92 commented:
@smichr  Thanks for your help ! Can we chat on https://gitter.im/sympy/sympy

smichr commented:
I'm there.

jayesh92 committed:
xrange -> range

smichr commented on sympy/core/compatibility.py:
```diff
@@ -92,7 +92,7 @@ def u_decode(x):
     exec_ = getattr(builtins, "exec")
 
-    xrange = range
+    range = range
```
This should remain as it was. If xrange is used it should be replaced with range in PY3 since there is no xrange in PY3.

smichr commented on sympy/core/compatibility.py:
```diff
@@ -136,7 +136,7 @@ def exec_(_code_, _globs_=None, _locs_=None):
             _locs_ = _globs_
         exec("exec _code_ in _globs_, _locs_")
 
-    xrange = xrange
```
both these lines can be deleted

jayesh92 committed:
Updated range(<anything>)[<anything>] to
list(range(<anything>))[<anything>], Found in only two files

resolved slice issue after changing xrange to range

xrange conflict resolved in examples directory

@smichr comments taken care

smichr commented:
Looking good so far. Your title is too long, however...please make it conform to [the guide](https://github.com/sympy/sympy/wiki/Development-workflow) -- see the section "Writing commit messages".

Also, consider making the title reflect what you actually did, because you didn't change *all* such cases. range(i)[0]-like things didn't need to be changed. What you are doing is working around xrange(foo)[slice] incompatibility.

jayesh92 commented:
ok I'll change the commit message. Can this be merged ?

smichr commented:
> Can this be merged ?

no, the merge button is grey. You will have to merge with master locally, fixing the conflicts.

jayesh92 commented:
@smichr are you resolving conflicts (as you've self assigned )  ?

smichr commented:
No, that "self-assign" is just a way for me to keep track of PRs I'm actively participating in.

jayesh92 commented:
ok @smichr I'll surely do it in a few days

asmeurer commented:
Others can't merge your branch or push commits to your pull requests for you, unfortunately. Only you can push to your fork.

smichr commented:
The conflicts are fresh in my mind. I'll commit a local branch if tests pass. Thanks for working on this, Jayesh and welcome to SymPy!

jayesh92 commented:
Thanks a lot @smichr for your help !
